react-app/components/FileManager/hooks.js

```diff
@@ -0,0 +1,257 @@
+import { useEffect, useState, useMemo, useCallback, useRef } from 'react';
+import { ChonkyActions, FileHelper } from 'chonky';
+import { v4 as uuidv4 } from 'uuid';
+
+import { convertFileMapToTree, findNodeByIdInTree } from './utils';
+
+export const useCustomFileMap = (prepareCustomFileMap) => {
+  const { baseFileMap, rootFolderId } = useMemo(prepareCustomFileMap, []);
+  const rootFolderIdFixed = rootFolderId;
+
+  const [fileMap, setFileMap] = useState(baseFileMap);
+  const [currentFolderId, setCurrentFolderId] = useState(rootFolderId);
+  const [assetsKeyToDelete, setAssetsKeyToDelete] = useState([]);
+
+  const resetFileMap = useCallback(() => {
+    setFileMap(baseFileMap);
+    setCurrentFolderId(rootFolderId);
+    setAssetsKeyToDelete([]);
+  }, [baseFileMap, rootFolderId]);
+
+  const currentFolderIdRef = useRef(currentFolderId);
+  useEffect(() => {
+    currentFolderIdRef.current = currentFolderId;
+  }, [currentFolderId]);
+
+  const deleteFiles = useCallback((files) => {
+    setFileMap((currentFileMap) => {
+      const newFileMap = { ...currentFileMap };
+      const fileMapToTree = convertFileMapToTree(rootFolderId, '', fileMap);
+
+      files.forEach((file) => {
+        const nodeTree = findNodeByIdInTree(fileMapToTree, file.id) || { metadata: {}, path: '' };
+        // If the node has metadata means that was saved previously
+        if (file.isSaved) {
+          const {
+            metadata: { asset_key }
+          } = nodeTree;
+          setAssetsKeyToDelete((prevPaths) => [...prevPaths, asset_key]);
+        }
+
+        if (file.parentId) {
+          const parent = newFileMap[file.parentId];
+          const newChildrenIds = parent.childrenIds.filter((id) => id !== file.id);
+
+          newFileMap[file.parentId] = {
+            ...parent,
+            childrenIds: newChildrenIds,
+            childrenCount: newChildrenIds.length
+          };
+        }
+      });
+
+      return newFileMap;
+    });
+  }, []);
+
+  const moveFiles = useCallback((files, source, destination) => {
+    setFileMap((currentFileMap) => {
+      const newFileMap = { ...currentFileMap };
+      const moveFileIds = new Set(files.map((f) => f.id));
+
+      const newSourceChildrenIds = source.childrenIds.filter((id) => !moveFileIds.has(id));
```

Comment:
[curious] why is it `newSource...`?

Reply:
Because we are moving the files (sources) to another folder and in general in functional programming we don't change the data (immutability) we replace it